### PR TITLE
fold: improve newline handling and test coverage

### PR DESCRIPTION
**Pull request by drocco007**
- refactor implementation for readability

- correct handling of files with no trailing newline and/or blank lines

### CODE REVIEW

File: src/uu/fold/src/fold.rs [hunk 1]
```diff
@@ -92,123 +91,173 @@ fn fold(filenames: Vec<String>, bytes: bool, spaces: bool, width: usize) {
...
-        if ch.is_whitespace() {
-            return Some(slice.chars().count() - (i + 1));
+/// Fold `file` to fit `width` (number of columns).
+///
+/// By default `fold` treats tab, backspace, and carriage return specially:
+/// tab characters count as 8 columns, backspace decreases the
+/// column count, and carriage return resets the column count to 0.
+///
+/// If `spaces` is `true`, attempt to break lines at whitespace boundaries.
+#[allow(unused_assignments)]
```
Comment (sylvestre) [inline src/uu/fold/src/fold.rs:171]: Why did you add this directive? thanks

Comment (drocco007) [reply src/uu/fold/src/fold.rs:171]: Hi @sylvestre. Without this directive, the compiler complains about the match arm handling `\t` on lines 222–228:

```rust
                '\t' => {
￼                    if col_count + 8 > width && !output.is_empty() {
￼                        emit_output!();
￼                    }
￼                    col_count += 8;
￼                    last_space = Some(char_count);
                }
```
https://github.com/uutils/coreutils/pull/2025/files#diff-c0cfac636e7701f77f3693e680acd57c6c41eac32ccf3cb63ea30738ebce8687R222-R227

Here is the compiler output without the directive:

```
$ cargo build -p uu_fold
warning: value assigned to `last_space` is never read
   --> src/uu/fold/src/fold.rs:198:13
    |
198 |             last_space = None;
    |             ^^^^^^^^^^
...
223 |                         emit_output!();
    |                         --------------- in this macro invocation
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?
```

Generally, we want `emit_output!()` to reset the last seen whitespace to `None` since we've just printed it out:

```
    output = "...foo bar baz"
last_space =            ^

<call emit_output!(), prints "...foo bar ">

    output = "baz"
last_space = None
```

Tab presents an edge case: if the tab pushes the output over `width`, we want to emit the output, but the tab character is now the last seen whitespace, so we overwrite the value set by the macro and point `last_space` at the tab.

Taking a fresh look at this, I think I see how to parameterize the macro to eliminate the need for the directive without triggering the compiler warning. Before I do that, though, does the `emit_output!()` macro improve the readability of the implementation? Is it an appropriate solution here?
